core/src/main/java/org/kiwix/kiwixmobile/core/webserver/ZimHostActivity.kt

```diff
@@ -0,0 +1,290 @@
+    presenter.loadBooks(sharedPreferenceUtil.hostedBooks)
+    if (ServerUtils.isServerStarted) {
+      ip = ServerUtils.getSocketAddress()
+      layoutServerStarted()
+    }
+  }
+
+  private fun saveHostedBooks(booksList: ArrayList<BooksOnDiskListItem>) {
+    sharedPreferenceUtil.hostedBooks =
+      booksList.filter { book -> book.isSelected && book is BookOnDisk && book.book.title != null }
+        .map { book -> (book as BookOnDisk).book.title }.toSet()
+  }
+
+  private fun layoutServerStarted() {
+    serverTextView.text = getString(R.string.server_started_message, ip)
+    startServerButton.text = getString(R.string.stop_server_label)
+    startServerButton.setBackgroundColor(resources.getColor(R.color.stopServer))
+    bookDelegate.selectionMode = SelectionMode.NORMAL
+    booksAdapter.notifyDataSetChanged()
+  }
+
+  private fun layoutServerStopped() {
+    serverTextView.text = getString(R.string.server_textview_default_message)
+    startServerButton.text = getString(R.string.start_server_label)
+    startServerButton.setBackgroundColor(resources.getColor(R.color.greenTick))
+    bookDelegate.selectionMode = SelectionMode.MULTI
+    booksAdapter.notifyDataSetChanged()
+  }
+
+  override fun onDestroy() {
+    super.onDestroy()
+    presenter.detachView()
+  }
+
+  private fun setUpToolbar() {
+    val toolbar = findViewById<Toolbar>(R.id.toolbar)
+    setSupportActionBar(toolbar)
+    supportActionBar!!.title = getString(R.string.menu_host_books)
+    supportActionBar!!.setHomeButtonEnabled(true)
+    supportActionBar!!.setDisplayHomeAsUpEnabled(true)
+
+    toolbar.setNavigationOnClickListener { onBackPressed() }
+  }
+
+  // Advice user to turn on hotspot manually for API<26
+  private fun startHotspotManuallyDialog() {
+
+    alertDialogShower.show(KiwixDialog.StartHotspotManually(),
+      {
+        launchTetheringSettingsScreen()
+        Unit
+      },
+      null,
+      {
+        progressDialog = ProgressDialog.show(
+          this,
+          getString(R.string.progress_dialog_starting_server), "",
+          true
+        )
+        startService(createHotspotIntent(ACTION_CHECK_IP_ADDRESS))
+        Unit
+      }
+    )
+  }
+
+  private fun createHotspotIntent(action: String): Intent =
+    Intent(this, HotspotService::class.java).setAction(action)
+
+  override fun onServerStarted(ipAddress: String) {
+    ip = ipAddress
+    layoutServerStarted()
+  }
+
+  override fun onServerStopped() {
+    layoutServerStopped()
+  }
+
+  override fun onServerFailedToStart() {
+    Toast.makeText(this, R.string.server_failed_toast_message, Toast.LENGTH_LONG).show()
+  }
+
+  private fun launchTetheringSettingsScreen() {
+    val intent = Intent(Intent.ACTION_MAIN, null)
+    intent.addCategory(Intent.CATEGORY_LAUNCHER)
+    val cn = ComponentName("com.android.settings", "com.android.settings.TetherSettings")
+    intent.component = cn
+    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
+    startActivity(intent)
```

Comment:
I have taken to writing these like
```
startActivity(
  Intent(/*whatever constructor*/).apply{
    addCategory(something)
    //etc
  }
)
```

Reply:
https://github.com/kiwix/kiwix-android/pull/1569/commits/713ba3047d560fa72fc9065d031ba9c8bd989fd0